### PR TITLE
Add proposal to introduce `toXml` and `fromXml` APIs

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject 

Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/2819
## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec

### CODE REVIEW

File: docs/proposals/toXml_fromXml_apis.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+## Summary
+The Ballerina Xmldata module doesn't have any API to convert the Ballerina Record/Map to XML and XML to map. 
+So, this proposal introduces a new `fromXml` API to convert the XML data to Record/Map and `toXml` API to convert the Record/Map to XML.
+
+## Goals
+Provide a way to perform conversions between XML and Record/Map.
+
+## Motivation
+At the moment, users have to write their own custom implementation to perform conversions between Ballerina records/Map to XML.
+Therefore, It would be easy for them if we provided APIs to convert.
```
Comment (MadhukaHarith92) [inline docs/proposals/toXml_fromXml_apis.md:18]:
```suggestion
Therefore, It would be easier for them if we provided APIs to convert.
```

File: docs/proposals/toXml_fromXml_apis.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+Note: This feature(fromRecord()) is also required by the connector team. When they write a connector for the SOAP backend service(e.g Netsuite Connector),
+they need to convert the Ballerina record value to XML payload. As mentioned in the summary, we don't have a way to do this conversion. 
+They are planning to remove their custom implementation and use `xmldata` module, once we have this feature. As per the [discussion](https://github.com/ballerina-platform/ballerina-standard-library/issues/2819),
```
Comment (MadhukaHarith92) [inline docs/proposals/toXml_fromXml_apis.md:22]: Do we need to include these internal details?

File: docs/proposals/toXml_fromXml_apis.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+public isolated function fromXml(xml xmlValue, typedesc<(map<anydata>)> returnType = <>) returns returnType|Error;
+```
+
+```ballerina
+# Converts a `Map` or `Record` representation to its XML representation.
+# The record has annotations to configure namespaces and attributes,  but others don't have these.
+#
+# + mapValue - The `Map` or `Record` representation source to be converted to XML
+# + return - XML representation of the given source if the source is
+# successfully converted or else an `xmldata:Error`
```
Comment (MadhukaHarith92) [inline docs/proposals/toXml_fromXml_apis.md:47]:
```suggestion
# successfully converted or else, an `xmldata:Error`
```

File: docs/proposals/toXml_fromXml_apis.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+public isolated function toXml(map<anydata> mapValue) returns xml|Error;
+```
+
+### Record Annotation Definitions:
+
+```ballerina
+# Defines the new name of the name.
```
Comment (MadhukaHarith92) [inline docs/proposals/toXml_fromXml_apis.md:54]: Is it "Defines the new name"?

File: docs/proposals/toXml_fromXml_apis.md (hunk 2 above)
Comment (daneshk) [reply docs/proposals/toXml_fromXml_apis.md:22]: Shall we reword this as follows,

When writing a Ballerina connector for a SOAP backend service like Netsuite connector, we need to convert the Ballerina record to XML and vice versa. In Ballerina, we don't have a standard way of converting Ballerina records to XML and followed our own custom logic in each connector. From this feature, we are trying to standardize the conversion and provide an easier way for conversions than reimplementing the logic.

File: docs/proposals/toXml_fromXml_apis.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+we are planning to introduce `fromXml` API for their requirement which is an improved API from `fromRecord`. 
+This will be used to do the conversion of the Ballerina record|map data to XML.
+
+## Description
+
+### APIs definition:
+
+```ballerina
+# Converts an XML to its `Map` or `Record` representation.
+# The namespaces and attributes will not be considered a special case.
```
Comment (daneshk) [inline docs/proposals/toXml_fromXml_apis.md:32]: what does this mean?

File: docs/proposals/toXml_fromXml_apis.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+# + xmlValue - The XML source to be converted to a given target type
+# + returnType - The `typedesc` of the `map<anydata>` that should be returned as a result
+# + return - The given target type representation of the given XML on success,
+# else returns an `xmldata:Error`
+public isolated function fromXml(xml xmlValue, typedesc<(map<anydata>)> returnType = <>) returns returnType|Error;
+```
+
+```ballerina
+# Converts a `Map` or `Record` representation to its XML representation.
+# The record has annotations to configure namespaces and attributes,  but others don't have these.
```
Comment (daneshk) [inline docs/proposals/toXml_fromXml_apis.md:43]:
```suggestion
# Attribute and Namespace annotations provided in the Ballerina record definition get into account when converting to XML.
```

File: docs/proposals/toXml_fromXml_apis.md (hunk 4 above)
Comment (daneshk) [reply docs/proposals/toXml_fromXml_apis.md:54]: `Defines the XML element name which matches the record field. The default value is the record field name.`

File: docs/proposals/toXml_fromXml_apis.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+# successfully converted or else an `xmldata:Error`
+public isolated function toXml(map<anydata> mapValue) returns xml|Error;
+```
+
+### Record Annotation Definitions:
+
+```ballerina
+# Defines the new name of the name.
+#
+# + value - The value of the new name
```
Comment (daneshk) [inline docs/proposals/toXml_fromXml_apis.md:56]:
```suggestion
# + value - The XML element name```

File: docs/proposals/toXml_fromXml_apis.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+public type NameConfig record {|
+    string value;
+|};
+```
+
+```ballerina
+# The annotation is used to specify the new name of the existing record name or field name according to the XML format.
+public annotation NameConfig Name on type, record field;
```
Comment (daneshk) [inline docs/proposals/toXml_fromXml_apis.md:64]:
```suggestion
# The Annotation is used to specify which XML element matches the record field.
public annotation NameConfig Name on type, record field;

```

File: docs/proposals/toXml_fromXml_apis.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+public annotation NamespaceConfig Namespace on type, record field;
+```
+
+```ballerina
+# The annotation is used to denote the field that is considered an attribute.
+public annotation Attribute on record field;
+```
+### Rules for performing conversions between Map record and XML
+
+The same rules of conversion between JSON and XML are used. But, this doesn't consider the attributes and namespaces a special case.
```
Comment (daneshk) [inline docs/proposals/toXml_fromXml_apis.md:84]: No. we need to specifically mention, that we are following OpenAPI specifications. Please reword this. provide reference link

Comment (kalaiyarasiganeshalingam) [reply docs/proposals/toXml_fromXml_apis.md:84]: Sorry, I have to remove the `record` word in the heading. 

This section has the rules of conversion between XML and record.
https://github.com/ballerina-platform/module-ballerina-xmldata/pull/432/files#diff-0af47c0e4375c73ef387e816b3b5270631c76efdfef7a3cb92288fa0d919f176R86

File: docs/proposals/toXml_fromXml_apis.md (hunk 5 above)
Comment (kalaiyarasiganeshalingam) [reply docs/proposals/toXml_fromXml_apis.md:32]: `fromXml` API considers namespace and attribute and treats them as special cases, but this API doesn't treat them as special cases.

**Review by daneshk — APPROVED** [review]
